fix: expect interface, not a specific implementation

blizzz opened the pull request:
* Resolves: https://github.com/owncloud/music/issues/1068

## Summary

- fixes a regression when deleting folders while music app was enabled, for a LazyRoot was passed to this method.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [X] Screenshots before/after for front-end changes
- [X] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

blizzz committed:
fix: expect interface, not a specific implementation

- fixes a regression when deleting folders while music app was enabled,
  for a LazyRoot was passed to this method.

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
fix: add typehine for IRootFolder

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz committed:
chore: ugly type juggling

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

come-nc commented on lib/private/Files/Cache/QuerySearchHelper.php:
```diff
@@ -196,23 +195,31 @@ public function searchInCaches(ISearchQuery $searchQuery, array $caches): array
 		$rootLength = strlen($path);
-		$mount = $root->getMount($path);
-		$storage = $mount->getStorage();
+		if (method_exists($root, 'getMount')) {
```
```suggestion
		$storage = null;
		if (method_exists($root, 'getMount')) {
```
Would feel more readable this way.

come-nc commented on lib/private/Files/Node/Root.php:
```diff
@@ -342,7 +343,7 @@ public function isShareable() {
 	 * @return Node
```
```suggestion
```

come-nc commented on lib/private/Files/FileInfo.php:
```diff
@@ -79,7 +80,7 @@ class FileInfo implements \OCP\Files\FileInfo, \ArrayAccess {
 	/**
 	 * @param string|boolean $path
-	 * @param Storage\Storage $storage
+	 * @param Storage\Storage|IStorage $storage
```
The return type of getStorage needs to be adapted as well, no?

come-nc commented on lib/private/Files/Cache/QuerySearchHelper.php:
```diff
@@ -196,23 +195,31 @@ public function searchInCaches(ISearchQuery $searchQuery, array $caches): array
 		$mountByMountPoint = ['' => $mount];
 
-		if (!$limitToHome) {
+		if (!$limitToHome && method_exists($root, 'getMountsIn')) {
```
Does it actually happen in the wild to not have these methods on `$root`?
If not they should be added to the interface, no?

blizzz replied:
I was haunted by `Folder` which has custom implementations… but this one looks rather safe indeed, at least on Github there does not seem to be a custom implemenation of it: https://github.com/search?q=org%3Anextcloud+%2Fimplements.*IRootFolder%2F&type=code

I'll add it to the interface then, i don't like handling and guessing on private classes at all.

JoshuaPettus commented:
Why is this slated for NC28?  It's a regression that effects NC26 and 27.

paulijar commented:
>  It's a regression that effects NC26 and 27.

And also NC25.

blizzz committed:
refactor: declare getMount() and getMountsIn() at IRootFolder

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
```
ERROR: UndefinedInterfaceMethod - lib/private/Files/Node/Folder.php:347:24 - Method OCP\Files\IRootFolder::createNode does not exist (see https://psalm.dev/181)
                return [$this->root->createNode(

```

Wow, how did this even work, ever? `createNode` is `protected`.

blizzz committed:
ci: pro forma check of existence of internal method

createNode() is protected and used by Folder, but being an internal-only
method it shall not be exposed in the Folder or IRootFolder interface.

Signed-off-by: Arthur Schiwon <blizzz@arthur-schiwon.de>

blizzz commented:
/backport to stable27

blizzz commented:
/backport to stable26

blizzz commented:
/backport to stable25

marcelklehr commented on lib/public/Files/IRootFolder.php:
```diff
@@ -55,4 +56,16 @@ public function getUserFolder($userId);
+	/**
+	 * @since 28.0.0
+	 */
+	public function getMount(string $mountPoint): IMountPoint;
```
I'm not sure what's the policy on adding to interfaces in backports

blizzz replied:
Ah, right, backports are not relevant anymore.

blizzz replied:
Correct to bring that up. Afaik the IRootFolder is not implemented by others, then additions to the API are not breaking.

blizzz replied:
Yes, only implemented in the server repo, only in `Root` and `LazyRoot`. https://github.com/search?q=org%3Anextcloud+%2Fimplements.*IRootFolder%2F&type=code

marcelklehr commented on lib/private/Files/FileInfo.php:
```diff
@@ -143,10 +144,7 @@ public function getPath() {
-	 * @return \OCP\Files\Storage
-	 */
-	public function getStorage() {
+	public function getStorage(): Storage\Storage|IStorage {
```
Will this clash with the public FileInfo interface?

blizzz replied:
It's legitimate to narrow down return types in implementations.

blizzz replied:
So it does not clash code wise, but you might argue it clashes with the phpdoc-type. But not really. The public interface documents `IStorage` to return, while the implementation returned `Storage` – which implements `IStorage` however..

`\OCP\Files\Storage`  is an interface deprecated since 9.0.0, and it is advised to use IStorage instead. 

Right now I am unsure about `Storage\Storage` (also implementing the deprecated interface…) as return type, I will double check this tomorrow.

blizzz replied:
World of pain. 

While `OCP\Files\Storage` was deprecated in 9.0.0, it got a new method in 12.0.0, which is missing in `OCP\Files\Storage\IStorage`.  Then there is private interface `OC\Files\Storage\Storage` which is then having three additional methods.

Adding the `IStorage` on my side is probably not necessary, as both `Storage` interfaces implement it. I'd also remove the return type hint to be on the safe side.

blizzz commented:
/backport to stable27

blizzz commented:
/backport to stable26

blizzz commented:
/backport to stable25

szaimen commented:
/backport to stable25
